src/Polly.Core/Simmy/Outcomes/OutcomeChaosStrategy.TResult.cs

```diff
@@ -0,0 +1,126 @@
+﻿using Polly.Telemetry;
+
+namespace Polly.Simmy.Outcomes;
+
+#pragma warning disable S3928 // Custom ArgumentNullException message
+
+internal class OutcomeChaosStrategy<T> : ReactiveMonkeyStrategy<T>
+{
+    private readonly ResilienceStrategyTelemetry _telemetry;
+
+    public OutcomeChaosStrategy(OutcomeStrategyOptions<Exception> options, ResilienceStrategyTelemetry telemetry)
+        : base(options)
+    {
+        Guard.NotNull(telemetry);
+
+        if (options.Outcome.Exception is null && options.OutcomeGenerator is null)
+        {
+            throw new ArgumentNullException(nameof(options.Outcome), "Either Outcome or OutcomeGenerator is required.");
+        }
+
+        _telemetry = telemetry;
+        Fault = options.Outcome;
+        OnFaultInjected = options.OnOutcomeInjected;
+        FaultGenerator = options.Outcome.Exception is not null ? (_) => new(options.Outcome) : options.OutcomeGenerator;
+    }
+
+    public OutcomeChaosStrategy(OutcomeStrategyOptions<T> options, ResilienceStrategyTelemetry telemetry)
+        : base(options)
+    {
+        Guard.NotNull(telemetry);
+
+        _telemetry = telemetry;
+        Outcome = options.Outcome;
+        OnOutcomeInjected = options.OnOutcomeInjected;
+        OutcomeGenerator = options.Outcome.HasResult ? (_) => new(options.Outcome) : options.OutcomeGenerator;
+    }
+
+    public Func<OnOutcomeInjectedArguments<T>, ValueTask>? OnOutcomeInjected { get; }
+
+    public Func<OnOutcomeInjectedArguments<Exception>, ValueTask>? OnFaultInjected { get; }
+
+    public Func<ResilienceContext, ValueTask<Outcome<T>>>? OutcomeGenerator { get; }
+
+    public Func<ResilienceContext, ValueTask<Outcome<Exception>>>? FaultGenerator { get; }
```

Comment:
Isn't outcome and fault generator basically the same? Both can inject an outcome and outcome can be either exception or result.

Reply:
In theory yes, however, I need to differentiate them since the way the exception is injected is a bit different than the result, for example, a null value is a valid result to inject, so in that scenario we never execute the user's callback, however, when the fault is null it means we do not inject any fault but we run the user's callback as normal.

That's why the class has two different constructors, and two different properties to differentiate a result from a fault, I see value in that in the sense that I think it's always better to be explicit as possible.

Reply:
> In theory yes, however, I need to differentiate them since the way the exception is injected is a bit different than the result, for example, a null value is a valid result to inject, so in that scenario we never execute the user's callback, however, when the fault is null it means we do not inject any fault but we run the user's callback as normal.

Can't all the above be expressed simply as nullable `Outcome<T>`?

- `null`: Nothing get's injected.
- `Outcome<T>`: That represents a result.
- `Outcome<T>`: That represents an exception.

Reply:
Unfortunately not because I have a constructor that defines explicitly that the `T` is an `Exception`

![image](https://github.com/App-vNext/Polly/assets/12032625/29d94a67-dbaa-430b-aa59-1c974f1f409a)

Plus I need to know what kind of outcome will be injected since exceptions are treated differently for the reasons I explained before.
![image](https://github.com/App-vNext/Polly/assets/12032625/11336dc3-a149-4545-bb00-e40a8a6e87de)

One thing I like about this approach is actually the fact that the strategy makes it clear when it comes to an exception as an outcome, I think it is useful for consumers to have that separation, thoughts?

Reply:
Food for though. 

Isn't injecting exceptions non-reactive strategy as you don't care about result? But there would have to be a new `FaultStrategyOptions` and some duplications, which make me prefer the current solution more. 

It just comes with this little gotchas such as `OutcomeStrategyOptions<Exception>`.

Reply:
> Isn't injecting exceptions non-reactive strategy as you don't care about result? 

@martintmk  Not necessarily, I can have a user delegate that returns something and I might want to inject faults into it, e.g:

```C#
var onFaultInjected = false;
var userDelegateExecuted = false;
var exceptionMessage = "Dummy exception";
var fault = new InvalidOperationException(exceptionMessage);

var options = new OutcomeStrategyOptions<Exception>
{
	InjectionRate = 0.6,
	Enabled = true,
	Randomizer = () => 0.5,
	Outcome = new Outcome<Exception>(fault),
	OnOutcomeInjected = args =>
	{
		args.Context.Should().NotBeNull();
		args.Context.CancellationToken.IsCancellationRequested.Should().BeFalse();
		onFaultInjected = true;
		return default;
	}
};

var sut = new ResiliencePipelineBuilder<HttpStatusCode>().AddChaosFault(options).Build();
await sut.Invoking(s => s.ExecuteAsync(async _ =>
{
	userDelegateExecuted = true;
	return await Task.FromResult(HttpStatusCode.OK);
}).AsTask())
	.Should()
	.ThrowAsync<InvalidOperationException>()
	.WithMessage(exceptionMessage);

userDelegateExecuted.Should().BeFalse();
onFaultInjected.Should().BeTrue();
```